### DISCUSSION TITLE
AttributeError: 'NoneType' object has no attribute 'CLOSED'

### DISCUSSION THREAD

**Discussion by Gigioliva** [encode/httpx#1951] [Potential Issue]
Hi,

I'm trying to fetch a simple HTML page but I get an error when the program ends.

```py
import httpx

client = httpx.Client()
client.get("https://www.google.com")
```

When the script ends I get:
```py
Exception ignored in: <function Client.__del__ at 0x10586fdc0>
Traceback (most recent call last):
  File "/my_path/.venv/lib/python3.9/site-packages/httpx/_client.py", line 1281, in __del__
  File "/my_path/.venv/lib/python3.9/site-packages/httpx/_client.py", line 1242, in close
  File "/my_path/.venv/lib/python3.9/site-packages/httpx/_transports/default.py", line 199, in close
  File "/my_path/.venv/lib/python3.9/site-packages/httpcore/_sync/connection_pool.py", line 299, in close
  File "/my_path/.venv/lib/python3.9/site-packages/httpcore/_sync/connection.py", line 155, in close
  File "/my_path/.venv/lib/python3.9/site-packages/httpcore/_sync/http11.py", line 198, in close
AttributeError: 'NoneType' object has no attribute 'CLOSED'
```

I use the latest version `0.21.1`.

Where is my mistake? Can you help me?

**Comment by shashankrnr32** [comment]
I am not sure know if this has been resolved, leaving my update here after I noticed the same issue with my development environment.

Some observations from my environment
1. Using an old virtual environment with all my debugging packages installed in the env was causing the issue
2. Created a new virtual environment and just installed `httpx`. Issue doesnt exist anymore.
3. The issue is also not reproducible with the dev setup of `httpx`.

Decided to narrow it down to some package in combination of `httpx` and **_to my surprise_**, installing `click` was causing this issue. I will try to debug more soon, but uninstalling `click` worked (atleast for me)

I currently do not have **any** idea as to why `click` is causing this issue. Hope to find out soon! 

This might not be the case for you! But thought I would leave my update here!

**Comment by tsotnikov** [comment]
I have run into the same issue trying to upgrade `httpx`, after trying several versions, it happened for `httpx >= 0.20.0`

It must be something related with garbage collection order. Very difficult to reproduce from scratch, but with the same setup it happened 100% of the times I tried.

If we take a look at the failing line:
`AttributeError: 'NoneType' object has no attribute 'CLOSED'`
The object is an enum class: https://github.com/encode/httpcore/blob/0.14.4/httpcore/_sync/http11.py#L198

I will also try to debug a bit more.

**Comment by lovelydinosaur** [comment]
This makes no sense to me at all?

How can this line...

```python
self._state = HTTPConnectionState.CLOSED
```

Fail with...

`AttributeError: 'NoneType' object has no attribute 'CLOSED'`

It's just a plain ol' `enum.IntEnum` class? It shouldn't ever be `None` unless it's getting monkey-patched somewhere, or something icky like that?

I can't reproduce this myself - is anyone able to give some more information on reproducing this. (How are you installing things? How are you running the script? How reliably can you reproduce this?)

**Reply by lovelydinosaur** [reply]
This reported issue looks the same... https://github.com/encode/httpx/issues/2024

What environment are you folks using to run/install `httpx`? (That particular case looks like anaconda on windows I think?)
It looks to me like some weird environment issue where `intenum` classes are being garbage collected and set to `None`?

**Reply by Gigioliva** [reply]
> What environment are you folks using to run/install httpx?

In my case, `poetry 1.1.9` as dependency manager on macOS 12

**Reply by lovelydinosaur** [reply]
@Gigioliva How reliably can you reproduce this?
Are you able to give me a complete walk-though of installing and running this in a clean environment, and seeing this error?

 &mdash; Thanks 🙏🏼

**Reply by Gigioliva** [reply]
To reproduce it, I created an env with the following setting:
```toml
[tool.poetry]
name = "test"
version = "0.1.0"
description = "Test"
authors = [""]

[tool.poetry.dependencies]
python = "~3.9"
httpx = "==0.21.1"

[build-system]
requires = ["poetry-core>=1.0.0"]
build-backend = "poetry.core.masonry.api"
```

And run this script as is (the client and the GET is executed with the file without functions):
```py
import httpx

client = httpx.Client()
client.get("https://www.google.com")
```

**Reply by lovelydinosaur** [reply]
Which version of Python are you using?

(And does it reproduce reliably for you?)

**Reply by Gigioliva** [reply]
> Which version of Python are you using?

3.9.7

> And does it reproduce reliably for you?

the error always occurs

**Comment by erdnaxeli** [comment]
I have an error, maybe similar, I am not sure.

```
Exception ignored in: <function AsyncClient.__del__ at 0x7f3418c3b550>
Traceback (most recent call last):
  File "/home/myapp/.local/lib/python3.9/site-packages/httpx/_client.py", line 2012, in __del__
TypeError: 'NoneType' object is not callable
```

The error comes from using `warnings`. If I just add if the same `__del__` method a line `print(warnings)` I get the same error on my new line.

I use httpx 0.21.3, on a debian-slim (10) docker with python 3.9.9. httpx is installed with the `--system` flag (that means without creating a virtual env).

**Reply by lovelydinosaur** [reply]
Related yes - similar [reported here in the Python bug tracker](https://bugs.python.org/issue29174), although that's listed as fixed, and I wasn't able to reproduce the issue there.

**Reply by lovelydinosaur** [reply]
Also eg. https://bytes.com/topic/python/answers/44368-__del__-logging

> It all depends. If your __del__ is being called via atexit() for application
cleanup, for example, logging may not be available to you because it has been
cleaned up beforehand.

Or a whole bunch of similar items if ya search for em... https://duckduckgo.com/?q=%22TypeError%3A+%27NoneType%27+object+is+not+callable%22+warnings+__del__+python&t=osx&ia=web

It looks to me like the expectations of what may be do-able with `__del__` on system exit are basically "all bets are off".

The redis client has a good example here...

* https://github.com/redis/redis-py/issues/1339#issuecomment-631007244
* https://github.com/redis/redis-py/commit/afb3b81dca308307a17409a0a97bb52b070716b8

They've opted to simply silence *any* exceptions that occur during `__del__`.

We've got a couple of options here...

* Keep the implicit close / warnings as a best-attempt.
* Drop the implicit close / warnings. Leave it up to Python to warn if there are unclosed resources. (Note that eg. `requests` doesn't perform an implicit close for you.)
* Rejig things so that we've got more reliable behaviour here. (Not sure how far this gets us. An explicit `._is_closed` state rather than relying on the enum being available. Assume that the `warnings` module may not be available, or perhaps keep a reference to it to prevent early deletion?)

**Comment by huwcbjones** [comment]
If you don't have http2 support (see this snippet from dnspython) and create a client, when `__del__` is called, an unhandled AttributeError is printed to stderrr.
https://github.com/rthalley/dnspython/blob/ed4be1c94a892832e078398d33b920f23e8ba05f/dns/query.py#L70-L76

It appears that the `h2` import error is fired before `self._transport = self._init_transport()` is called, therefore the attribute is not present when `__del__` is called.

Run this example
```py
import httpx

try:
    # See if http2 support is available.
    with httpx.Client(http2=True):
        pass
except Exception:
    pass
```
Outputs something like
```
Exception ignored in: <function Client.__del__ at 0x7fb043532790>
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/httpx/_client.py", line 1139, in __del__
    self.close()
  File "/usr/lib/python3/dist-packages/httpx/_client.py", line 1111, in close
    self._transport.close()
AttributeError: 'Client' object has no attribute '_transport'
```
